### PR TITLE
Don't hold the GIL when calling anything in core

### PR DESCRIPTION
**Pull request by soltanmm**

### CODE REVIEW

File: src/python/grpcio/grpc/_cython/_cygrpc/channel.pyx.pxi [hunk 1]
```diff
@@ -45,12 +46,14 @@ cdef class Channel:
...
       raise TypeError("expected target to be str or bytes")
+    c_target = target
     if channel_credentials is None:
-      self.c_channel = grpc_insecure_channel_create(target, c_arguments,
-                                                         NULL)
+      with nogil:
+        self.c_channel = grpc_insecure_channel_create(c_target, c_arguments,
+                                                      NULL)
     else:
       self.c_channel = grpc_secure_channel_create(
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_cython/_cygrpc/channel.pyx.pxi]: Missing `with nogil:` in this `else:` branch?

File: src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi [hunk 2]
```diff
@@ -82,8 +83,10 @@ cdef class CompletionQueue:
   def poll(self, Timespec deadline=None):
     # We name this 'poll' to avoid problems with CPython's expectations for
     # 'special' methods (like next and __next__).
-    cdef gpr_timespec c_deadline = gpr_inf_future(
-        GPR_CLOCK_REALTIME)
+    cdef gpr_timespec c_deadline
+    with nogil:
+      c_deadline = gpr_inf_future(
+          GPR_CLOCK_REALTIME)
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi]: Unnecessary line-wrap?

File: src/python/grpcio/grpc/_cython/_cygrpc/records.pyx.pxi [hunk 3]
```diff
@@ -240,20 +252,27 @@ cdef class ByteBuffer:
...
-      while grpc_byte_buffer_reader_next(&reader, &data_slice):
-        data_slice_pointer = gpr_slice_start_ptr(data_slice)
-        data_slice_length = gpr_slice_length(data_slice)
-        result += (<char *>data_slice_pointer)[:data_slice_length]
-      grpc_byte_buffer_reader_destroy(&reader)
+      with nogil:
+        while grpc_byte_buffer_reader_next(&reader, &data_slice):
+          data_slice_pointer = gpr_slice_start_ptr(data_slice)
+          data_slice_length = gpr_slice_length(data_slice)
+          with gil:
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_cython/_cygrpc/records.pyx.pxi]: `with gil:` inside `with nogil:`: not sure if impressed or terrified.

File: src/python/grpcio/grpc/_cython/_cygrpc/records.pyx.pxi [hunk 4]
```diff
@@ -240,20 +252,27 @@ cdef class ByteBuffer:
     cdef size_t data_slice_length
     cdef void *data_slice_pointer
     if self.c_byte_buffer != NULL:
-      grpc_byte_buffer_reader_init(&reader, self.c_byte_buffer)
+      with nogil:
+        grpc_byte_buffer_reader_init(&reader, self.c_byte_buffer)
       result = b""
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_cython/_cygrpc/records.pyx.pxi]: Outside the scope of this change, but isn't this quadratic in time? I suppose we don't care until we have performance metrics telling us it's a problem?

File: src/python/grpcio/grpc/_cython/_cygrpc/records.pyx.pxi [hunk 5]
```diff
@@ -240,20 +252,27 @@ cdef class ByteBuffer:
...
       return result
     else:
       return None
 
   def __len__(self):
+    cdef size_t result
     if self.c_byte_buffer != NULL:
-      return grpc_byte_buffer_length(self.c_byte_buffer)
+      with nogil:
+        result = grpc_byte_buffer_length(self.c_byte_buffer)
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_cython/_cygrpc/records.pyx.pxi]: Can we not have a return statement inside a `with nogil:`?

File: src/python/grpcio/grpc/_cython/_cygrpc/records.pyx.pxi [hunk 6]
```diff
@@ -319,14 +339,16 @@ cdef class ChannelArgs:
       if not isinstance(arg, ChannelArg):
         raise TypeError("expected list of ChannelArg")
     self.c_args.arguments_length = len(self.args)
-    self.c_args.arguments = <grpc_arg *>gpr_malloc(
-        self.c_args.arguments_length*sizeof(grpc_arg)
-    )
+    with nogil:
+      self.c_args.arguments = <grpc_arg *>gpr_malloc(
+          self.c_args.arguments_length*sizeof(grpc_arg)
+      )
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_cython/_cygrpc/records.pyx.pxi]: Move this to the end of the previous line?

There's not yet a cyformat to match pyformat, right? :-P

File: src/python/grpcio/grpc/_cython/_cygrpc/records.pyx.pxi [hunk 7]
```diff
@@ -648,8 +674,9 @@ cdef class Operations:
       if not isinstance(operation, Operation):
         raise TypeError("expected operations to be iterable of Operation")
     self.c_nops = len(self.operations)
-    self.c_ops = <grpc_op *>gpr_malloc(
-        sizeof(grpc_op)*self.c_nops)
+    with nogil:
+      self.c_ops = <grpc_op *>gpr_malloc(
+          sizeof(grpc_op)*self.c_nops)
```
Comment (nathanielmanistaatgoogle) [inline src/python/grpcio/grpc/_cython/_cygrpc/records.pyx.pxi]: Unnecessary line-wrap?

**Comment by nathanielmanistaatgoogle** [comment]
Looks mostly good; I think all my comments felt minor.

Even if just on your local machine, is this something we want to test five times, twenty times, a hundred times, or more?

File: src/python/grpcio/grpc/_cython/_cygrpc/records.pyx.pxi (hunk 3 above)
Comment (soltanmm) [inline src/python/grpcio/grpc/_cython/_cygrpc/records.pyx.pxi]: It was either that or adding yet another temporary. :-/

File: src/python/grpcio/grpc/_cython/_cygrpc/records.pyx.pxi (hunk 4 above)
Comment (soltanmm) [inline src/python/grpcio/grpc/_cython/_cygrpc/records.pyx.pxi]: Yes, but AFAIK it tends to loop only once. We can get back to it at some point.

File: src/python/grpcio/grpc/_cython/_cygrpc/records.pyx.pxi (hunk 5 above)
Comment (soltanmm) [inline src/python/grpcio/grpc/_cython/_cygrpc/records.pyx.pxi]: If it's returning a Python object that needs to be constructed with the GIL locked, we can't. Cython's full of magic but not that much magic.

**Comment by soltanmm** [comment]
Repeating offline statements: my local machine's been running this on repeat for the past two hours. :-) Running with a thousand iterations, and I guess we'll know by tomorrow.

**Comment by soltanmm** [comment]
Current Python `mac,opt` failure on Jenkins is a transient build failure.

**Comment by nathanielmanistaatgoogle** [comment]
`'_ssl.c:574: The handshake operation timed out'`; I agree.

**Comment by nathanielmanistaatgoogle** [comment]
How'd the thousand iterations turn out?

**Comment by soltanmm** [comment]
No freezes having to do with the change, as far as I can tell. There was the already known pipe-buffer-deadlock issue we discussed earlier this afternoon, but that was it.

**Comment by nathanielmanistaatgoogle** [comment]
No further comments; this looks good.

In accordance with _Keeping The Build Green Going Forward_ this will require a rerun of the basic tests, right?

**Comment by soltanmm** [comment]
Ayep. I can't press the button from the DMV, though. -.-'

**Comment by nathanielmanistaatgoogle** [comment]
Python is green.

**Comment by jtattermusch** [comment]
Merging this as it should further decrease python flakiness.

**Pull request merged**